File: src/test/java/kitchenpos/products/application/InMemoryProductRepository.java
package kitchenpos.products.application;

import kitchenpos.products.domain.ProductRecord;
import kitchenpos.products.domain.ProductRepository;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.UUID;

public class InMemoryProductRepository implements ProductRepository {
    private final Map<UUID, ProductRecord> products = new HashMap<>();

    @Override
    public ProductRecord save(final ProductRecord product) {
        products.put(product.getId(), product);
        return product;
    }

    @Override
    public Optional<ProductRecord> findById(final UUID id) {
        return Optional.ofNullable(products.get(id));
    }

    @Override
    public List<ProductRecord> findAll() {
        return new ArrayList<>(products.values());
    }

    @Override
    public List<ProductRecord> findAllByIdIn(final List<UUID> ids) {
        return products.values()
            .stream()
            .filter(product -> ids.contains(product.getId()))
            .toList();
    }
}

File: src/test/java/kitchenpos/products/application/InMemoryProductRepository.kt
package kitchenpos.products.application

import kitchenpos.products.domain.Product
import kitchenpos.products.domain.ProductRepository
import java.util.*

class InMemoryProductRepository : ProductRepository {
    private val products: MutableMap<UUID, Product> = mutableMapOf()

    override fun save(product: Product): Product {
        products[product.id] = product
        return product
    }

    override fun findById(id: UUID?): Optional<Product> {
        return Optional.ofNullable(products[id])
    }

    override fun findAll(): List<Product> {
        return products.values.toList()
    }

    override fun findAllByIdIn(ids: MutableList<UUID>?): MutableList<Product> {
        return products.filterKeys { ids?.contains(it) ?: false }.values.toMutableList()
    }
}
